lib/Doctrine/KeyValueStore/Storage/AmazonDynamoDbStorage.php

```diff
@@ -0,0 +1,311 @@
+     *
+     * @return string The default key name.
+     */
+    public function getDefaultKeyName()
+    {
+        return $this->defaultKeyName;
+    }
+
+    /**
+     * Sets a key name for a specific table.
+     *
+     * @param $table string The name of the table.
+     * @param $key string The name of the string.
+     *
+     * @throws \InvalidArgumentException When the key or table name is invalid.
+     */
+    public function setKeyForTable($table, $key)
+    {
+        $this->validateTableName($table);
+        $this->validateKeyName($key);
+        $this->tableKeys[$table] = $key;
+    }
+
+    /**
+     * Retrieves a specific name for a key for a given table. The default is returned if this table does not have
+     * an actual override.
+     *
+     * @param string $tableName The name of the table.
+     *
+     * @return string
+     */
+    public function getKeyNameForTable($tableName)
+    {
+        return isset($this->tableKeys[$tableName]) ?
+            $this->tableKeys[$tableName] :
+            $this->defaultKeyName;
+    }
+
+    /**
+     * Prepares a key to be in a valid format for lookups for DynamoDB. If passing an array, that means that the key
+     * is the name of the key and the value is the actual value for the lookup.
+     *
+     * @param string $storageName
+     *
+     * @return array The key in DynamoDB format.
+     */
+    protected function prepareKey($storageName, $key)
+    {
+        if (is_array($key)) {
+            $keyValue = reset($key);
+            $keyName = key($key);
+        } else {
+            $keyValue = $key;
+            $keyName = $this->getKeyNameForTable($storageName);
+        }
+
+        return $this->marshaler->marshalItem([$keyName => $keyValue]);
+    }
+
+    /**
+     * Determine if the storage supports updating only a subset of properties,
+     * or if all properties have to be set, even if only a subset of properties
+     * changed.
+     *
+     * @return bool
+     */
+    public function supportsPartialUpdates()
+    {
+        //This is not true, but partial updates are too complicated given the available interface,
+        //meaning, the abstraction is insufficiently flexible enough to support this type of action.
+        return false;
+    }
+
+    /**
+     * Does this storage support composite primary keys?
+     *
+     * @return bool
+     */
+    public function supportsCompositePrimaryKeys()
+    {
+        return true;
+    }
+
+    /**
+     * Does this storage require composite primary keys?
+     *
+     * @return bool
+     */
+    public function requiresCompositePrimaryKeys()
+    {
+        return false;
+    }
+
+    /**
+     * Insert data into the storage key specified.
+     *
+     * @param string       $storageName
+     * @param array|string $key
+     * @param array        $data
+     */
+    public function insert($storageName, $key, array $data)
+    {
+        $this->client->putItem([
+            'TableName' => $storageName,
```

Comment:
Consider using class constants for these - makes it easier to refactor and to understand/document the meaning